### PR TITLE
MESOS: update known issues

### PR DESCRIPTION
**Pull request by jdef**
Better communicate known issues
- xref https://github.com/mesosphere/kubernetes-mesos/issues/604

### CODE REVIEW

File: contrib/mesos/docs/issues.md [hunk 1]
```diff
@@ -1,34 +1,120 @@
-## Known Issues
+## DCOS Package Known Issues
```
Comment (karlkfi) [inline contrib/mesos/docs/issues.md]: It's super confusing have this as the page title when not all of the issues are DCOS specific.

It'd be ideal to just have a "Known Issues" page title, folowed bya  table of contents that shows that there are general issues followed by dcos-specific issues. Thent he DCOS section can say that all of the general issues apply, and these additional issues also apply.

File: contrib/mesos/docs/issues.md [hunk 2]
```diff
@@ -1,34 +1,120 @@
...
 
-### Pod Placement
+These known issues apply specifically to the Kubernetes-Mesos DCOS package available from https://github.com/mesosphere/multiverse.
+
+### Kubectl
+
+The following `kubectl` and `dcos kubectl` commands are not supported:
```
Comment (karlkfi) [inline contrib/mesos/docs/issues.md]: should say why they're not supported, maybe link to ticket?

File: contrib/mesos/docs/issues.md [hunk 3]
```diff
@@ -1,34 +1,120 @@
...
+It is **strongly** recommended that all of the Kubernetes-Mesos executors are destroyed before upgrading the Kubernetes-Mesos scheduler component:
+- destroy all daemon controllers running in the cluster, across all namespaces
+- destroy all replication controllers running in the cluster, across all namespaces
+- destroy all pods running in the cluster, across all namespaces
+- invoke the "kamikaze" debug endpoint on the scheduler (e.g. `curl http://10.2.0.5:10251/debug/actions/kamikaze`) to terminate all executors
+
+Not following the above steps prior to upgrading the scheduler can result in a cluster wherein pods fail to launch on some or all nodes - indefinitely.
+This issue is being tracked here: https://github.com/mesosphere/kubernetes-mesos/issues/572.
+
+### Ports
```
Comment (karlkfi) [inline contrib/mesos/docs/issues.md]: Pod Ports?

File: contrib/mesos/docs/issues.md [hunk 4]
```diff
@@ -1,34 +1,120 @@
...
+  To obtain the same behavior with the Kubernetes-Mesos scheduler pods must be assigned a label of `k8s.mesosphere.io/portMapping` with the value `fixed`
+  (see [#527](https://github.com/mesosphere/kubernetes-mesos/issues/527)).
+
+### Pods
+
+#### Pod Updates
+
+Once a task has been launched for a given pod, Kubernetes-Mesos is blind to any updates applied to the pod state (other than for forced, or graceful deletion).
+
+#### Pod Placement
```
Comment (karlkfi) [inline contrib/mesos/docs/issues.md]: do we want to link to the new scheduling doc here?

**Comment by karlkfi** [comment]
The ports stuff is kinda complicated. I'd really like to be able to easily answer from the user's perspective "What do I need to do about it?" for services, pods, and mesos slaves.  The services section is really clear with the what and the how, without getting too deep into the why.  I want something similar for pod config and mesos-slave config.  The why should still be in the doc, but you don't want to get to the ports section and see the why first, see a giant list of long bullet points and not know what to do about it.

File: contrib/mesos/docs/issues.md (hunk 4 above)
Comment (jdef) [inline contrib/mesos/docs/issues.md]: done

File: contrib/mesos/docs/issues.md [hunk 5]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
+That scheduler will pair each resource offer with a pod that needs to be placed in the cluster.
+Once paired, a task is launched to instantiate the pod.
+The used resources will be marked as consumed, the remaining resources are "returned" to the cluster and the scheduler will wait for the next resource offer from the master... and the cycle repeats itself.
+This likely limits the scheduling throughput observable in a single-node cluster.
+
+We are currently running some benchmarks on the scheduling algorithm to establish some baselines, and are definitely thinking about ways to increase scheduling throughput- including scheduling multiple pods per offer.
+
+#### Runtime Configuration
+
+- mesos: `--offer_timeout` : Duration of time before an offer is rescinded from a framework.
```
Comment (sttts) [inline contrib/mesos/docs/issues.md]: --offer-timeout? Didn't check, but I guess it's with a dash as well.

File: contrib/mesos/docs/issues.md [hunk 6]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
+
+### Kubectl
+
+The following `kubectl` and `dcos kubectl` commands are not yet supported:
+
+- exec (see [#356](https://github.com/mesosphere/kubernetes-mesos/issues/356))
+- logs (see [#587](https://github.com/mesosphere/kubernetes-mesos/issues/587))
+- port-forward
+- proxy
+- version (see [#632](https://github.com/mesosphere/kubernetes-mesos/issues/632))
```
Comment (sttts) [inline contrib/mesos/docs/issues.md]: version works now.

File: contrib/mesos/docs/issues.md [hunk 7]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
+
+If no matching resource offer can be found for a pod then that pod is put into a backoff queue.
+Once the backoff period expires the pod is re-added to the scheduling queue.
+The backoff period may be truncated by the arrival of an offer with matching resources.
+This is an event-based design and there is no polling.
+
+#### Debugging
+
+Debugging can be challenging.
+The best experience I have had operationally is when all of the relevant logs are collected into a single tool (Splunk, or an ELK stack) in a manner such that it is trivial to search for something along the lines of a task-id or pod-id during cluster debugging sessions.
```
Comment (sttts) [inline contrib/mesos/docs/issues.md]: third person, not "I"

File: contrib/mesos/docs/issues.md [hunk 8]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
+Statements in this section regarding the "scheduler" pertain specifically to the Kubernetes-Mesos scheduler, unless otherwise noted.
 
-There are two work-arounds to this problem:
-* Restart the framework and it should terminate the orphaned tasks.
-* Adjust the value of `executor_shutdown_grace_period` to something greater than 3 seconds.
+Some factors that influence when pods are scheduled by k8s-mesos:
+- availability of a resource offer that "fits" the pod (mesos master/slave);
+- scheduler *backoff* (to avoid busy-looping) during pod scheduling (k8s-mesos scheduler)
+
+We have tried to mitigate the second item by cancelling the backoff period if an offer arrives that fits a pod-in-waiting.
```
Comment (sttts) [inline contrib/mesos/docs/issues.md]: Why "we", third person.

File: contrib/mesos/docs/issues.md [hunk 9]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
+
+The slave on each host announces offers to Mesos periodically.
+In a single node cluster only a single slave process is advertising resources to the master.
+The master will pass those along to the scheduler, at some interval and level of 'fairness' determined by mesos.
+That scheduler will pair each resource offer with a pod that needs to be placed in the cluster.
+Once paired, a task is launched to instantiate the pod.
+The used resources will be marked as consumed, the remaining resources are "returned" to the cluster and the scheduler will wait for the next resource offer from the master... and the cycle repeats itself.
+This likely limits the scheduling throughput observable in a single-node cluster.
+
+We are currently running some benchmarks on the scheduling algorithm to establish some baselines, and are definitely thinking about ways to increase scheduling throughput- including scheduling multiple pods per offer.
```
Comment (sttts) [inline contrib/mesos/docs/issues.md]: Why "we", third person.

File: contrib/mesos/docs/issues.md [hunk 10]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
+#### Runtime Configuration
+
+- mesos: `--offer_timeout` : Duration of time before an offer is rescinded from a framework.
+  This helps fairness when running frameworks that hold on to offers, or frameworks that accidentally drop offers.
+  ([via](http://mesos.apache.org/documentation/latest/configuration/))
+- k8s-mesos `--scheduler-config` : An ini-style configuration file with low-level scheduler settings.
+  See `offer-ttl`, `initial-pod-backoff`, and `max-pod-backoff`.
+  ([via](https://github.com/kubernetes/kubernetes/blob/master/contrib/mesos/pkg/scheduler/config/config.go))
+
+What's not configurable, but perhaps should be, are the mesos "filters" that we include when declining offers that we cannot match to pods within the config'd `offer-ttl` (see https://github.com/apache/mesos/blob/0.25.0/include/mesos/mesos.proto#L1165). I think we hardcode the `refuse_seconds` value to 5s.
```
Comment (sttts) [inline contrib/mesos/docs/issues.md]: Why "we", third person.

File: contrib/mesos/docs/issues.md [hunk 11]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
+We should probably expose that via the scheduler fine tuning mechanism.
```
Comment (sttts) [inline contrib/mesos/docs/issues.md]: Why "we", third person.

File: contrib/mesos/docs/issues.md [hunk 12]
```diff
@@ -1,34 +1,118 @@
...
+
+Upgrading your Kubernetes-Mesos cluster is currently "YMMV" status.
+One particularly nasty problem exists with respect to expressing executor (kubelet and kube-proxy) process configuration via command line flags.
+It is **strongly** recommended that all of the Kubernetes-Mesos executors are destroyed before upgrading the Kubernetes-Mesos scheduler component:
+- destroy all daemon controllers running in the cluster, across all namespaces
+- destroy all replication controllers running in the cluster, across all namespaces
+- destroy all pods running in the cluster, across all namespaces
+- invoke the "kamikaze" debug endpoint on the scheduler (e.g. `curl http://10.2.0.5:10251/debug/actions/kamikaze`) to terminate all executors
+
+Not following the above steps prior to upgrading the scheduler can result in a cluster wherein pods fail to launch on some or all nodes - indefinitely.
```
Comment (sttts) [inline contrib/mesos/docs/issues.md]: In fact they don't fail AFAIK. Don't we decline offers with incompatible executor id group? Then we should never see a TASK_ERROR.

File: contrib/mesos/docs/issues.md [hunk 13]
```diff
@@ -1,34 +1,118 @@
...
+* [General Known Issues](#general-known-issues)
+* [DCOS Package Known Issues](#dcos-package-known-issues), in addendum to the above.
+
+## General Known Issues
+
+These known issues apply to all builds of Kubernetes-Mesos.
+
+### Upgrades
+
+Upgrading your Kubernetes-Mesos cluster is currently "YMMV" status.
```
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: What is a "YMMV" status? Even when knowing this acronym I'd still be puzzled ;-) I'd be very explicit here, i.e.
"Upgrading a Kubernetes-Mesos cluster is currently unsupported."

File: contrib/mesos/docs/issues.md [hunk 14]
```diff
@@ -1,34 +1,118 @@
...
+One particularly nasty problem exists with respect to expressing executor (kubelet and kube-proxy) process configuration via command line flags.
```
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: "nasty" is a subjective adjective. I'd me more objective like "One known problem exists with respect ..."

File: contrib/mesos/docs/issues.md [hunk 15]
```diff
@@ -1,34 +1,118 @@
...
+- destroy all replication controllers running in the cluster, across all namespaces
+- destroy all pods running in the cluster, across all namespaces
+- invoke the "kamikaze" debug endpoint on the scheduler (e.g. `curl http://10.2.0.5:10251/debug/actions/kamikaze`) to terminate all executors
+
+Not following the above steps prior to upgrading the scheduler can result in a cluster wherein pods fail to launch on some or all nodes - indefinitely.
+This issue is being tracked here: https://github.com/mesosphere/kubernetes-mesos/issues/572.
+
+### Port Specifications
+
+In order for Pods (replicated, or otherwise) to be scheduled on the cluster, it is strongly recommended that:
```
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: s/Pods/pods

File: contrib/mesos/docs/issues.md [hunk 16]
```diff
@@ -1,34 +1,118 @@
...
+Upgrading your Kubernetes-Mesos cluster is currently "YMMV" status.
+One particularly nasty problem exists with respect to expressing executor (kubelet and kube-proxy) process configuration via command line flags.
+It is **strongly** recommended that all of the Kubernetes-Mesos executors are destroyed before upgrading the Kubernetes-Mesos scheduler component:
+- destroy all daemon controllers running in the cluster, across all namespaces
+- destroy all replication controllers running in the cluster, across all namespaces
+- destroy all pods running in the cluster, across all namespaces
+- invoke the "kamikaze" debug endpoint on the scheduler (e.g. `curl http://10.2.0.5:10251/debug/actions/kamikaze`) to terminate all executors
+
+Not following the above steps prior to upgrading the scheduler can result in a cluster wherein pods fail to launch on some or all nodes - indefinitely.
+This issue is being tracked here: https://github.com/mesosphere/kubernetes-mesos/issues/572.
```
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: :+1: (signals we're working on it)

File: contrib/mesos/docs/issues.md [hunk 17]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
-* Restart the framework and it should terminate the orphaned tasks.
-* Adjust the value of `executor_shutdown_grace_period` to something greater than 3 seconds.
+Some factors that influence when pods are scheduled by k8s-mesos:
+- availability of a resource offer that "fits" the pod (mesos master/slave);
+- scheduler *backoff* (to avoid busy-looping) during pod scheduling (k8s-mesos scheduler)
+
+We have tried to mitigate the second item by cancelling the backoff period if an offer arrives that fits a pod-in-waiting.
+However, there is nothing that the scheduler can do if there are no resources available in the cluster.
+
+That said, the current scheduling algorithm is somewhat naive: it makes **no attempts to pack multiple pods into a single offer**.
```
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: s/somewhat//

File: contrib/mesos/docs/issues.md (hunk 5 above)
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: `--offer_timeout` is correct (see http://mesos.apache.org/documentation/latest/configuration/)

File: contrib/mesos/docs/issues.md (hunk 10 above)
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: Yeah and the "I think ..." is strange. Let's check whether we hardcoded it and make it a fact here.

File: contrib/mesos/docs/issues.md [hunk 18]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
+#### Backoff
+
+If no matching resource offer can be found for a pod then that pod is put into a backoff queue.
+Once the backoff period expires the pod is re-added to the scheduling queue.
+The backoff period may be truncated by the arrival of an offer with matching resources.
+This is an event-based design and there is no polling.
+
+#### Debugging
+
+Debugging can be challenging.
```
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: I'd leave out this sentence, it is subjective.
Or make it more objective, i.e. "Debugging can be achieved with one of the following methods:"

File: contrib/mesos/docs/issues.md (hunk 7 above)
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: I'd also leave out "The best experience I have had operationally" or replace it with "Good insights can be achieved when all of the relevant logs ..."

File: contrib/mesos/docs/issues.md [hunk 19]
```diff
@@ -37,27 +121,93 @@ Kubernetes clusters assume a custom Docker networking configuration that assigns
...
+The best experience I have had operationally is when all of the relevant logs are collected into a single tool (Splunk, or an ELK stack) in a manner such that it is trivial to search for something along the lines of a task-id or pod-id during cluster debugging sessions.
+
+The scheduler also offers some `/debug` API endpoints that may be useful:
```
Comment (s-urbaniak) [inline contrib/mesos/docs/issues.md]: s/some//

**Comment by s-urbaniak** [comment]
@jdef left some nit comments, thanks for providing this!

**Comment by sttts** [comment]
lgtm. please squash

**Comment by jdef** [comment]
squished

**Comment by jdef** [comment]
thanks @wojtek-t I didn't realize that label was there!

**Comment by wojtek-t** [comment]
@jdef - we are using "e2e-not-required" label for all PRs that cannot affect tests that are running as part of PR builder tests (in particular docs-only changes, etc.)

**Pull request merged**